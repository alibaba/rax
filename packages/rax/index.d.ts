import Rax from './lib/index';

export = Rax;
