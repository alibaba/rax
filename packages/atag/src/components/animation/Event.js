import { requestFrame } from './helpers';
import Pan from './Pan';
import Orientation from './Orientation';

export default class Event {
  constructor({ origin, type, callback }) {
    this.origin = origin;
    this.type = type;
    this.callback = callback;

    if (this.type) {
      requestFrame(this._listen.bind(this));
    }
  }

  _getElement() {
    if (!this.origin) {
      return null;
    }
    if (this.origin === '@window') {
      return window;
    } else if (this.origin === '@document') {
      return document;
    } else if (this.origin === '@body') {
      return document.body;
    } else {
      return document.querySelector(`#${this.origin}`);
    }
  }

  /**
   * 监听事件
   */
  _listen() {
    this.element = this._getElement();
    // auto start animation with no type or type is timing
    if (this.type === 'timing') {
      this.callback();
    } else if (this.type === 'pan') {
      this.pan = new Pan(this.element);
      this.pan.on('pan', this._onPan);
      this.pan.on('panstart', this._onPanStart);
      this.pan.on('panend', this._onPanEnd);
    } else if (this.type === 'orientation') {
      this.orientation = new Orientation(this.callback);
    } else {
      this.element && this.element.addEventListener(this.type, this.callback, false);
    }
  }

  /**
   * get node translateXY
   * @param {object} node element
   */
  getComputedTranslateXY = (node) => {
    const transformArray = [];
    if (!window.getComputedStyle) return transformArray;
    const style = getComputedStyle(node);
    const transform = style.transform || style['-webkit-transform'];

    let matrix = transform.match(/^matrix3d\((.+)\)$/);
    if (matrix) {
      const matrixs = matrix[1].split(', ');
      // 3d
      transformArray.push(parseFloat(matrixs[12]));
      transformArray.push(parseFloat(matrixs[13]));
      return transformArray;
    } else {
      // 2d
      matrix = transform.match(/^matrix\((.+)\)$/);
      if (matrix) {
        const matrixs = matrix[1].split(', ');
        transformArray.push(parseFloat(matrixs[4]));
        transformArray.push(parseFloat(matrixs[5]));
      }
      return transformArray;
    }
  }

  /**
   * get transform data
   * @param {boolean} start is start status
   * @return {object} transform data
   *   {number} startTranslateX touch start translateX
   *   {number} startTranslateY touch start translateY
   *   {number} translateX touch translateX
   *   {number} translateY touch translateY
   */
  getTransformData(start = false) {
    const transform = {};
    // event type is pan, add transform data
    const translate = this.getComputedTranslateXY(this.element);
    const translateX = translate[0] || 0;
    const translateY = translate[1] || 0;
    transform.translateX = translateX;
    transform.translateY = translateY;
    // start transform data
    if (start) {
      this.element.startTranslateX = translateX;
      this.element.startTranslateY = translateY;
    }
    transform.startTranslateX = this.element.startTranslateX;
    transform.startTranslateY = this.element.startTranslateY;

    return transform;
  }

  _onPan = (e) => {
    this.callback({
      x: e.deltaX,
      y: e.deltaY,
      clientX: e.touches[0].clientX,
      clientY: e.touches[0].clientY,
      ...this.getTransformData()
    });
  }

  _onPanStart = (e) => {
    this.callback({
      x: 0,
      y: 0,
      clientX: e.touches[0].clientX,
      clientY: e.touches[0].clientY,
      state: 'start',
      ...this.getTransformData(true)
    });
  }

  _onPanEnd = (e) => {
    this.callback({
      x: e.deltaX,
      y: e.deltaY,
      state: 'end',
      ...this.getTransformData()
    });
  }

  /**
   * 销毁事件
   */
  destroy() {
    if (this.type === 'pan') {
      if (this.pan) {
        this.pan.off('pan', this._onPan);
        this.pan.off('panstart', this._onPanStart);
        this.pan.off('panend', this._onPanEnd);
        this.pan.destroy();
      }
    } else if (this.type === 'orientation') {
      this.orientation && this.orientation.destroy();
    } else {
      this.element && this.element.removeEventListener(this.type, this.callback);
    }
  }
}