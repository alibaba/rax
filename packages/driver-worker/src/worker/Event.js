export default class Event {
  constructor(type, options = {}) {
    Object.assign(this, options); // extend event.

    this.defaultPrevented = false;
    this.bubbles = !!options.bubbles;
    this.cancelable = !!options.cancelable;
    this.type = type.toLowerCase();
  }
  stopPropagation() {
    this.bubbles = false;
  }
  stopImmediatePropagation() {
    this.bubbles = false;
    this._end = true;
  }
  preventDefault() {
    if (this.cancelable) {
      this.defaultPrevented = true;
    } else {
      // Calling preventDefault in uncancelable event should produce errors,
      // but in chrome and safari, which not throw or console errors, follow behaviors here.
    }
  }
}
