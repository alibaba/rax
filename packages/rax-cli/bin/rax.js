#!/usr/bin/env node

// Update notifications
const updateNotifier = require('update-notifier');
const pkg = require('../package.json');

updateNotifier({pkg: pkg}).notify();

// Check node version
const chalk = require('chalk');
const semver = require('semver');

if (!semver.satisfies(process.version, '>=6')) {
  const message = 'You are currently running Node.js ' +
    chalk.red(process.version) + '.\n' +
    '\n' +
    'Rax runs on Node 6.0 or newer. There are several ways to ' +
    'upgrade Node.js depending on your preference.\n' +
    '\n' +
    'nvm:       nvm install node && nvm alias default node\n' +
    'Homebrew:  brew install node\n' +
    'Installer: download the Mac .pkg from https://nodejs.org/\n';

  console.log(message);
  process.exit(1);
}

const fs = require('fs');
const path = require('path');
const execSync = require('child_process').execSync;
const spawn = require('cross-spawn');
const inquirer = require('inquirer');
const chalk = require('chalk');
const cli = require('../src/');
const argv = require('minimist')(process.argv.slice(2));

const RAX_PACKAGE_JSON_PATH = path.resolve(
  process.cwd(),
  'node_modules',
  'rax',
  'package.json'
);

checkForVersionArgument();

// minimist api
const commands = argv._;

if (commands.length === 0) {
  console.error(
    'You did not pass any commands, did you mean to run `rax init`?'
  );
  process.exit(1);
}

switch (commands[0]) {
  case 'init':
    if (!commands[1]) {
      console.error(
        'Usage: rax init <ProjectName> [--verbose]'
      );
      process.exit(1);
    } else {
      init(commands[1], argv.verbose, argv.version);
    }
    break;
  default:
    console.error(
      'Command `%s` unrecognized.',
      commands[0]
    );
    process.exit(1);
    break;
}

function init(name, verbose) {
  Promise.resolve(fs.existsSync(name))
    .then(function(dirExists) {
      if (dirExists) {
        return createAfterConfirmation(name, verbose);
      } else {
        return;
      }
    })
    .then(function() {
      return askProjectInformaction(name, verbose);
    })
    .then(function(answers) {
      return createProject(name, verbose, answers);
    })
    .catch(function(err) {
      console.error('Error occured', err.stack);
      process.exit(1);
    });
}

function createAfterConfirmation(name, verbose) {
  var property = {
    type: 'confirm',
    name: 'continueWhileDirectoryExists',
    message: 'Directory ' + name + ' already exists. Continue?',
    default: false
  };

  return inquirer.prompt(property).then(function(answers) {
    if (answers && answers.continueWhileDirectoryExists) {
      return true;
    } else {
      console.log('Project initialization canceled.');
      process.exit(1);
    }
  });
}

function askProjectInformaction(name, verbose) {
  var questions = [
    {
      type: 'input',
      name: 'projectName',
      message: 'What\'s your project name?',
      default: name
    },
    {
      type: 'input',
      name: 'author',
      message: 'What\'s author\'s name?',
      default: 'rax'
    },
    {
      type: 'confirm',
      name: 'autoInstallModules',
      message: 'Do you want to install dependences automatically after initialization?',
      default: 'y'
    }
  ];
  return inquirer.prompt(questions);
}

function createProject(name, verbose, userAnswers) {
  var pkgManager = shouldUseYarn() ? 'yarn' : 'npm';
  var root = path.resolve(name);
  var projectName = userAnswers.projectName;
  var projectAuthor = userAnswers.author;
  var autoInstallModules = userAnswers.autoInstallModules;

  console.log(
    'Creating a new Rax project in',
    root
  );

  if (!fs.existsSync(root)) {
    fs.mkdirSync(root);
  }
  process.chdir(root);

  cli.init({
    root: root,
    directoryName: name,
    projectName: projectName,
    projectAuthor: projectAuthor,
    verbose: verbose,
  }).then(function(directory) {
    if (autoInstallModules) {
      return install(directory, verbose);
    } else {
      return false;
    }
  }).then(function(isAutoInstalled) {
    console.log(chalk.white.bold('To run your app:'));
    console.log(chalk.white('   cd ' + projectName));
    if (!isAutoInstalled) {
      console.log(chalk.white('   ' + (pkgManager === 'npm' ? 'npm install' : 'yarn')));
    }
    console.log(chalk.white('   ' + pkgManager + ' run start'));
  });
}

function checkForVersionArgument() {
  if (argv._.length === 0 && (argv.v || argv.version)) {
    console.log('rax-cli: ' + require('../package.json').version);
    try {
      console.log('rax: ' + require(RAX_PACKAGE_JSON_PATH).version);
    } catch (e) {
      console.log('rax: n/a - not inside a Rax project directory');
    }
    process.exit();
  }
}

function shouldUseYarn() {
  try {
    execSync('yarn --version', {stdio: 'ignore'});
    return true;
  } catch (e) {
    return false;
  }
}

/**
 * run npm/yarn install
 * @param directory - the cwd directory
 * @param verbose - enable verbose mode
 * @returns {Promise}
 */
function install(directory, verbose) {
  console.log(chalk.white.bold('Install dependencies:'));

  var pkgManager = shouldUseYarn() ? 'yarn' : 'npm';
  var args = ['install'];
  if (verbose) {
    args.push('--verbose');
  }

  return new Promise(function(resolve) {
    var proc = spawn(pkgManager, args, {stdio: 'inherit', cwd: directory});

    proc.on('close', function(code) {
      if (code !== 0) {
        console.error('`' + pkgManager + ' install` failed');
        resolve(false);
      } else {
        resolve(true);
      }
    });
  });
}
