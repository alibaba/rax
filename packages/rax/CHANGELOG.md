## CHANGELOG

### 1.2.2

- fix: `lib/compat` bundle error path

### 1.2.1

- refactor: change output result format, avoid mixing `IIFE` and `CJS`
- chore: avoid throw error when update unmounted component
- fix: add error message to `getDerivedStateFromError` hook
