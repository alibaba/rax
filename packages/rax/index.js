if (process.env.NODE_ENV === 'production') {
  module.exports = require('./dist/rax.min.js');
} else {
  module.exports = require('./dist/rax.js');
}
