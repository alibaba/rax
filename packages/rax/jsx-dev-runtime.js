export { Fragment, jsxDEV } from './index';
