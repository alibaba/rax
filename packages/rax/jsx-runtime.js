export { Fragment, jsx, jsxs, jsxDEV } from './index';
