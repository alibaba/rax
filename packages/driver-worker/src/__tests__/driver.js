import { render, createElement } from 'rax';
import Event from '../worker/Event';
import createDriver from '../index';

describe('driver-worker', () => {
  it('should be drived by rax', (done) => {
    const addEventListener = () => {};
    const postMessage = (message) => {
      expect(message).toMatchSnapshot();
      done();
    };
    const driver = createDriver({ addEventListener, postMessage });
    render(createElement('view'), null, { driver });
  });

  it('should works with dom operation', (done) => {
    const addEventListener = (eventName, handler) => {
      // console.log('addEventListener', eventName, handler);
    };
    const postMessage = (message) => {
      expect(message).toMatchSnapshot();
      done();
    };
    const driver = createDriver({ addEventListener, postMessage });

    // Create body
    const body = driver.createBody();
    // Create element
    const container = driver.createElement({ type: 'view', props: { count: 0 } });
    // Append child
    body.appendChild(container);
    // Create comment
    const comment = driver.createComment('empty');
    driver.appendChild(comment, container);

    // removeChild
    const toBeRemoved = driver.createElement({ type: 'be-remove' });
    body.appendChild(toBeRemoved);
    driver.removeChild(toBeRemoved, body);

    // replaceChild
    const toBeReplaced = driver.createElement({ type: 'be-replace' });
    const toReplace = driver.createElement({ type: 'replaced' });
    body.appendChild(toBeRemoved);
    driver.replaceChild(toReplace, toBeReplaced, body);

    // insertAfter
    const toBeInsertAfter = driver.createElement({ type: 'be-insert-after' });
    driver.insertAfter(toBeInsertAfter, comment, body);

    // insertBefore
    const toBeInsertBefore = driver.createElement({ type: 'be-insert-before' });
    driver.insertBefore(toBeInsertBefore, comment, body);
  });

  it('should works with device width and viewport width', () => {
    let sendMessageToWorker;
    const addEventListener = (event, handler) => {
      sendMessageToWorker = handler;
    };
    const postMessage = (message) => {};
    const driver = createDriver({ addEventListener, postMessage });

    // Default device width be undefined
    expect(driver.getDeviceWidth()).toBeUndefined();

    // Mock init event.
    sendMessageToWorker({
      data: {
        type: 'init',
        width: 100,
      },
    });
    expect(driver.getDeviceWidth()).toEqual(100);

    // Set device width
    driver.setDeviceWidth(200);
    expect(driver.getDeviceWidth()).toEqual(200);

    // Set viewport width
    driver.setViewportWidth(100);
    expect(driver.getViewportWidth()).toEqual(100);
  });


  it('shuold work with event listener API', (done) => {
    const addEventListener = (event, handler) => { };
    const postMessage = (message) => {};
    const driver = createDriver({ addEventListener, postMessage });

    const body = driver.createBody();
    const onMove = (event) => {
      throw new Error('should not fire move event.');
    };
    const el = driver.createElement({
      type: 'view',
      props: {
        onMove,
        onClick: (event) => done(),
      },
    });
    body.appendChild(el);

    driver.removeEventListener(el, 'move', onMove);
    el.dispatchEvent({
      type: 'move',
      target: el,
    });

    el.dispatchEvent({
      type: 'click',
      target: el,
    });
  });

  it('shuold work with remove all event listeners', (done) => {
    const addEventListener = (event, handler) => { };
    const postMessage = (message) => {};
    const driver = createDriver({ addEventListener, postMessage });

    const body = driver.createBody();
    const toThrow = (event) => {
      throw new Error('should not fire move event.');
    };
    const el = driver.createElement({
      type: 'view',
      props: {
        onMove: toThrow,
        onClick: toThrow,
      },
    });
    body.appendChild(el);

    // Remove all event listeners will not fire anymore.
    driver.removeAllEventListeners(el);
    el.dispatchEvent({
      type: 'move',
      target: el,
    });
    el.dispatchEvent({
      type: 'click',
      target: el,
    });
    done();
  });

  it('should work with attributes', () => {
    const addEventListener = (event, handler) => { };
    const postMessage = (message) => {};
    const driver = createDriver({ addEventListener, postMessage });

    const body = driver.createBody();
    const el = driver.createElement({
      type: 'view',
    });
    body.appendChild(el);

    driver.setAttribute(el, 'key', 'value');
    expect(el.getAttribute('key')).toEqual('value');
  });

  it('should work with setNativeProps', (done) => {
    const addEventListener = (event, handler) => { };
    const postMessage = (message) => {
      expect(message).toMatchSnapshot();
      done();
    };
    const driver = createDriver({ addEventListener, postMessage });

    const body = driver.createBody();
    const el = driver.createElement({
      type: 'view',
    });
    driver.setNativeProps(el, {
      onClick: () => {},
      foo: 'bar',
      number: 0,
    });
    body.appendChild(el);
  });

  it('sanitize', () => {
    const addEventListener = (event, handler) => { };
    const postMessage = (message) => {};
    const driver = createDriver({ addEventListener, postMessage });
    const el = driver.createElement({ type: 'foo' });
    el.setAttribute('foo', 'bar');
    el.style = { color: 'red' };

    expect(driver.sanitize(el, 'addedNodes')).toMatchSnapshot();
    expect(driver.sanitize([el, el], 'addedNodes')).toMatchSnapshot();
  });

  it('convertTouchTarget', () => {
    const addEventListener = (event, handler) => { };
    const postMessage = (message) => {};
    const driver = createDriver({ addEventListener, postMessage });
    const event = new Event('touchstart');
    driver.nodesMap.set('0', 'FOO');
    const touch = {
      $$id: '0',
    };
    event.touches = [touch];
    event.changedTouches = [touch];
    driver.convertTouchTarget(event);

    expect(event).toMatchSnapshot();
  });

  it('appendChild', () => {
    const addEventListener = (event, handler) => { };
    const postMessage = (message) => {};
    const driver = createDriver({ addEventListener, postMessage });
    const parent = driver.createElement({ type: 'parent' });
    const son = driver.createElement({ type: 'son' });
    driver.appendChild(son, parent);
    expect(son.parentNode).toEqual(parent);
  });

  it('removeChild', () => {
    const addEventListener = (event, handler) => { };
    const postMessage = (message) => {};
    const driver = createDriver({ addEventListener, postMessage });
    const parent = driver.createElement({ type: 'parent' });
    const son = driver.createElement({ type: 'son' });
    parent.appendChild(son);
    driver.removeChild(son, parent);
    expect(parent.childNodes.length).toEqual(0);
  });
});
