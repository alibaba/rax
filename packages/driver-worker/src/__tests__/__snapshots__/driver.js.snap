// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`driver-worker convertTouchTarget 1`] = `
Event {
  "bubbles": false,
  "cancelable": false,
  "changedTouches": Array [
    Object {
      "target": "FOO",
    },
  ],
  "defaultPrevented": false,
  "touches": Array [
    Object {
      "target": "FOO",
    },
  ],
  "type": "touchstart",
}
`;

exports[`driver-worker sanitize 1`] = `
Object {
  "$$id": "1",
  "attributes": Array [
    Object {
      "name": "foo",
      "value": "bar",
    },
  ],
  "nodeName": "FOO",
  "nodeType": 1,
  "style": Object {
    "color": "red",
  },
}
`;

exports[`driver-worker sanitize 2`] = `
Array [
  Object {
    "$$id": "1",
    "attributes": Array [
      Object {
        "name": "foo",
        "value": "bar",
      },
    ],
    "nodeName": "FOO",
    "nodeType": 1,
    "style": Object {
      "color": "red",
    },
  },
  Object {
    "$$id": "1",
    "attributes": Array [
      Object {
        "name": "foo",
        "value": "bar",
      },
    ],
    "nodeName": "FOO",
    "nodeType": 1,
    "style": Object {
      "color": "red",
    },
  },
]
`;

exports[`driver-worker should be drived by rax 1`] = `
Object {
  "mutations": Array [
    Object {
      "addedNodes": Array [
        Object {
          "$$id": "1",
          "nodeName": "VIEW",
          "nodeType": 1,
        },
      ],
      "target": Object {
        "$$id": "2",
        "nodeName": "BODY",
      },
      "type": "childList",
    },
  ],
  "type": "MutationRecord",
}
`;

exports[`driver-worker should work with setNativeProps 1`] = `
Object {
  "mutations": Array [
    Object {
      "addedNodes": Array [
        Object {
          "$$id": "1",
          "attributes": Array [
            Object {
              "name": "foo",
              "value": "bar",
            },
            Object {
              "name": "number",
              "value": 0,
            },
          ],
          "events": Array [
            "click",
          ],
          "nodeName": "VIEW",
          "nodeType": 1,
        },
      ],
      "target": Object {
        "$$id": "2",
        "nodeName": "BODY",
      },
      "type": "childList",
    },
  ],
  "type": "MutationRecord",
}
`;

exports[`driver-worker should works with dom operation 1`] = `
Object {
  "mutations": Array [
    Object {
      "addedNodes": Array [
        Object {
          "$$id": "6",
          "attributes": Array [
            Object {
              "name": "count",
              "value": 0,
            },
          ],
          "nodeName": "VIEW",
          "nodeType": 1,
        },
      ],
      "target": Object {
        "$$id": "3",
        "nodeName": "BODY",
      },
      "type": "childList",
    },
    Object {
      "addedNodes": Array [
        Object {
          "$$id": "1",
          "data": "empty",
          "nodeType": 8,
        },
      ],
      "target": Object {
        "$$id": "6",
      },
      "type": "childList",
    },
    Object {
      "addedNodes": Array [
        Object {
          "$$id": "5",
          "nodeName": "BE-REMOVE",
          "nodeType": 1,
        },
      ],
      "target": Object {
        "$$id": "3",
        "nodeName": "BODY",
      },
      "type": "childList",
    },
    Object {
      "removedNodes": Array [
        Object {
          "$$id": "5",
        },
      ],
      "target": Object {
        "$$id": "3",
        "nodeName": "BODY",
      },
      "type": "childList",
    },
    Object {
      "removedNodes": Array [
        Object {
          "$$id": "5",
        },
      ],
      "target": Object {
        "$$id": "3",
        "nodeName": "BODY",
      },
      "type": "childList",
    },
    Object {
      "addedNodes": Array [
        Object {
          "$$id": "5",
          "nodeName": "BE-REMOVE",
          "nodeType": 1,
        },
      ],
      "target": Object {
        "$$id": "3",
        "nodeName": "BODY",
      },
      "type": "childList",
    },
    Object {
      "addedNodes": Array [
        Object {
          "$$id": "4",
          "nodeName": "BE-INSERT-AFTER",
          "nodeType": 1,
        },
      ],
      "target": Object {
        "$$id": "3",
        "nodeName": "BODY",
      },
      "type": "childList",
    },
    Object {
      "addedNodes": Array [
        Object {
          "$$id": "2",
          "nodeName": "BE-INSERT-BEFORE",
          "nodeType": 1,
        },
      ],
      "nextSibling": Object {
        "$$id": "1",
      },
      "target": Object {
        "$$id": "3",
        "nodeName": "BODY",
      },
      "type": "childList",
    },
  ],
  "type": "MutationRecord",
}
`;
